### PR TITLE
Added Junit Tests for service/api module

### PR DESCRIPTION
**Pull request by sonja-ct**
Moved class KapuaExceptionWithoutBundle from test to main folder

Added Junit tests for KapuaEntityCloneException class
Added Junit tests for KapuaEntityExistsException
Added Junit tests for KapuaEntityUniquenessException class
Added Junit tests for KapuaException class
Added Junit tests for KapuaExceptionWithoutBundle class
Added Junit tests for KapuaIllegalAccessException class
Added Junit tests for KapuaIllegalNullArgumentException class
Added Junit tests for KapuaIllegalStateException class
Added Junit tests for KapuaMaxNumberOfItemsReachedException class
Added Junit tests for KapuaOptimisticLockingException class
Added Junit tests for KapuaUnauthenticatedException class

Signed-off-by: Sonja <sonja.matic@comtrade.com>

**Related Issue**
/

**Description of the solution adopted**
/

**Screenshots**
/

**Any side note on the changes made**
/

### CODE REVIEW

File: service/api/src/test/java/org/eclipse/kapua/ExceptionMessageUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    public void exceptionMessageUtilsTest() throws NoSuchMethodException, IllegalAccessException, InvocationTargetException, InstantiationException {
+        Constructor<ExceptionMessageUtils> exceptionMessageUtils = ExceptionMessageUtils.class.getDeclaredConstructor();
+        assertTrue(Modifier.isPrivate(exceptionMessageUtils.getModifiers()));
+        exceptionMessageUtils.setAccessible(true);
+        exceptionMessageUtils.newInstance();
+    }
+
+    @Test
+    public void getLocalizedMessageTest() {
+        for (int i = 0; i < resourceBundleName.length; i++) {
```
Comment (LeoNerdoG) [inline service/api/src/test/java/org/eclipse/kapua/ExceptionMessageUtilsTest.java]: Please use enhanced for loop (String string: resourceBundleName)

File: service/api/src/test/java/org/eclipse/kapua/ExceptionMessageUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            assertEquals("Expected and actual values should be the same.", kapuaGenericMessageWithObject[i], ExceptionMessageUtils.getLocalizedMessage(resourceBundleName[i], locale, kapuaErrorCode, objectList));
+            assertEquals("Expected and actual values should be the same.", kapuaGenericMessage[i], ExceptionMessageUtils.getLocalizedMessage(resourceBundleName[i], locale, kapuaErrorCode, null));
```
Comment (LeoNerdoG) [inline service/api/src/test/java/org/eclipse/kapua/ExceptionMessageUtilsTest.java]: You are not covering line 57 of the tested class, please take a look.

File: service/api/src/test/java/org/eclipse/kapua/ExceptionMessageUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void getLocalizedMessageNullStringTest() {
+        ExceptionMessageUtils.getLocalizedMessage(null, locale, kapuaErrorCode, objectList);
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void getLocalizedMessageNullLocaleTest() {
+        for (int i = 0; i < resourceBundleName.length; i++) {
```
Comment (LeoNerdoG) [inline service/api/src/test/java/org/eclipse/kapua/ExceptionMessageUtilsTest.java]: Please use enhanced for loop (String string: resourceBundleName)

File: service/api/src/test/java/org/eclipse/kapua/KapuaExceptionTest.java [hunk 4]
```diff
@@ -12,37 +12,293 @@
...
+                assertEquals("NullPointerException expected.", new NullPointerException().toString(), e.toString());
+            }
+            try {
+                kapuaException.getLocalizedMessage();
+                fail("NullPointerException expected.");
+            } catch (Exception e) {
+                assertEquals("NullPointerException expected.", new NullPointerException().toString(), e.toString());
+            }
+        }
+    }
```
Comment (LeoNerdoG) [inline service/api/src/test/java/org/eclipse/kapua/KapuaExceptionTest.java:196]: No tests for: KapuaException kapuaException = new KapuaException(null, null, null);?

File: service/api/src/test/java/org/eclipse/kapua/KapuaExceptionTest.java [hunk 5]
```diff
@@ -12,37 +12,293 @@
...
-        KapuaExceptionWithoutBundle exceptionWithoutBundle = new KapuaExceptionWithoutBundle(KapuaErrorCodes.ILLEGAL_STATE, "param1", "param2");
-        Assert.assertEquals("Error: param1, param2", exceptionWithoutBundle.getMessage());
+    public void internalErrorCauseTest() {
+        String message = "Message";
+        Throwable nullThrowable = null;
+        Throwable[] throwables = {new Throwable(message), new Throwable()};
+        String[] expectedMessage = {"An internal error occurred: " + message + ".", "An internal error occurred: " + throwables[1] + "."};
+        String[] arguments = {"Message", "java.lang.Throwable"};
+
+        for (int i = 0; i < throwables.length; i++) {
```
Comment (LeoNerdoG) [inline service/api/src/test/java/org/eclipse/kapua/KapuaExceptionTest.java:224]: Enhanced for loop? 
        for (Throwable throwableParameter: throwables) {

File: service/api/src/test/java/org/eclipse/kapua/KapuaExceptionWithoutBundleTest.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Eurotech and/or its affiliates and others
```
Comment (LeoNerdoG) [inline service/api/src/test/java/org/eclipse/kapua/KapuaExceptionWithoutBundleTest.java:2]: Pleas move class "KapuaExceptionBundle" from test directory to src directory!

**Review by LeoNerdoG — CHANGES_REQUESTED** [review]
Hey @sonja-ct,
I have reviewed this PR and there are some minor changes to be made.

File: service/api/src/test/java/org/eclipse/kapua/KapuaExceptionTest.java (hunk 4 above)
Comment (sonja-ct) [reply service/api/src/test/java/org/eclipse/kapua/KapuaExceptionTest.java:196]: It is covered in
  for (Throwable throwable : throwables) {
            KapuaException kapuaException = new KapuaException(null, throwable, null);

File: service/api/src/test/java/org/eclipse/kapua/KapuaExceptionTest.java (hunk 5 above)
Comment (sonja-ct) [reply service/api/src/test/java/org/eclipse/kapua/KapuaExceptionTest.java:224]: I used index i to get values from different arrays in the same position

**Review by LeoNerdoG — APPROVED** [review]

File: service/api/src/test/java/org/eclipse/kapua/ExceptionMessageUtilsTest.java (hunk 1 above)
Comment (sonja-ct) [reply service/api/src/test/java/org/eclipse/kapua/ExceptionMessageUtilsTest.java]: I moved this class to PR#3070 due to number of lines.  But here I used index i to get values from different arrays in the same position.

**Comment by LeoNerdoG** [comment]
Hey @Coduz,
I have reviewed this PR and it looks OK. Travis is green and coverage increased, can you please take a look at this and merge when you can? 
Thanks!

**Review by Coduz — APPROVED** [review]

**Pull request merged**